Charon/filetypes/CuraPackage.py

```diff
@@ -0,0 +1,124 @@
+    PLUGIN_IGNORED_FILES = {r"__pycache__", r"\.qmlc", r"\.pyc"}
+    REQUIRED_METADATA_FIELDS = {"package_id", "author/author_id"}
+
+    _global_metadata_file = "/package.json"
+    _metadata_relationship_type = "http://schemas.ultimaker.org/package/2018/relationships/curapackage_metadata"
+    mime_type = "application/x-curapackage"
+
+    # File aliases for quick and easy access.
+    _aliases = OrderedDict([
+        (r"/materials", "/files/resources/materials"),
+        (r"/qualities", "/files/resources/qualities"),
+        (r"/definitions", "/files/resources/definitions"),
+        (r"/plugins", "/files/plugins")
+    ])
+
+    ##  Gets a list of paths to material files in the package.
+    def getMaterials(self) -> List[str]:
+        return self.listPaths("/materials")
+
+    ##  Add a new material file.
+    #   \param material_data The data of the material file in bytes.
+    #   \param package_filename The location to write the data to inside the package.
+    def addMaterial(self, material_data: bytes, package_filename: str) -> None:
+        material_path_alias = "/materials"
+        self._ensureRelationExists(virtual_path=material_path_alias, relation_type="material", origin="/package.json")
+        self._writeToAlias(material_path_alias, package_filename, material_data)
+
+    ##  Gets a list of paths to quality files in the package.
+    def getQualities(self) -> List[str]:
+        return self.listPaths("/qualities")
+
+    ##  Add a new quality file.
+    #   \param: quality_data The data of the quality file in bytes.
+    #   \param package_filename The location to write the data to inside the package.
+    def addQuality(self, quality_data: bytes, package_filename: str) -> None:
+        quality_path_alias = "/qualities"
+        self._ensureRelationExists(virtual_path=quality_path_alias, relation_type="quality", origin="/package.json")
+        self._writeToAlias(quality_path_alias, package_filename, quality_data)
+
+    ##  Gets a list of paths to machine definition files in the package.
+    def getMachines(self) -> List[str]:
+        return self.listPaths("/definitions")
+
+    ##  Add a new machine definition file.
+    #   \param machine_data The data of the machine definition file in bytes.
+    #   \param package_filename The location to write the data to inside the package.
+    def addMachine(self, machine_data: bytes, package_filename: str) -> None:
+        machine_path_alias = "/definitions"
+        self._ensureRelationExists(virtual_path=machine_path_alias, relation_type="definition", origin="/package.json")
+        self._writeToAlias(machine_path_alias, package_filename, machine_data)
+
+    ##  Gets a list of paths to plugins in the package.
+    def getPlugins(self) -> List[str]:
+        return self.listPaths("/plugins")
+
+    ##  Add a new plugin.
+    #   \param plugin_data The ZIP file containing the plugin files converted to bytes.
```

Comment:
Nitpick; Zip data from the file as bytes (It's not being converted, that's simply what you get if you read it as binary)